src/ray/core_worker/shared_actor_info_accessor.cc

```diff
@@ -0,0 +1,104 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/core_worker/shared_actor_info_accessor.h"
+
+namespace ray {
+
+SharedActorInfoAccessor::SharedActorInfoAccessor(
+    std::shared_ptr<gcs::GcsClient> gcs_client)
+    : gcs_client_(gcs_client) {}
+
+Status SharedActorInfoAccessor::AsyncSubscribe(
```

Comment:
Why not just making existing ActorInfoAccessor thread safe instead of implementing a new one?

Reply:
It's not about thread safety. Due to the implementation of GCS client, we can subscribe to each actor only once. If we try to subscribe to the same actor multiple times, the worker process will crash with the error `Caller attempted a duplicate subscribe or unsubscribe to channel ...`. But in the multiple `CoreWorker` instances scenario, because each `CoreWorker` instance has an `ActorManager` instance, it's very likely that multiple `ActorManager` instances subscribe to the same actor.

The only purpose of `SharedActorInfoAccessor` is to avoid duplicate subscriptions by making sure we invoke `gcs_client_->Actors().AsyncSubscribe(...)` only once for each actor and re-route the callbacks.

Reply:
It's possible that we allow multiple subscriptions to the same actor while making sure multiple callbacks are invoked correctly by modifying the implementation of GCS client. But I don't have this knowledge. @WangTaoTheTonic may have a better idea if this is easy to do.

Reply:
Hmm I see. I prefer the latter approach because it seems to be much more straightforward (for maintainers) to modify gcs client & calling it it can be shared by multiple threads. It is kind of hard to reason about that we should use this shared acessor when we already have gcs_client_->Actor()->Subscribe. 

Is this PR a blocker for the metrics agent port discovery PR? 

Btw, I think thread-safety on NodeInfoAccessor is still needed  if gcs_client is shared because `gcs_client->Nodes()->Get` will access the node cache which can be modified by other threads.

Reply:
> Hmm I see. I prefer the latter approach because it seems to be much more straightforward (for maintainers) to modify gcs client & calling it it can be shared by multiple threads. It is kind of hard to reason about that we should use this shared acessor when we already have gcs_client_->Actor()->Subscribe.
> 
> Is this PR a blocker for the metrics agent port discovery PR?
> 
> Btw, I think thread-safety on NodeInfoAccessor is still needed if gcs_client is shared because `gcs_client->Nodes()->Get` will access the node cache which can be modified by other threads.

No, this PR does not block the metrics agent port discovery. That PR fails on Windows CI (passes all cases on other platforms). I don't have a Windows dev environment, I am trying to fix it as soon as possible.